### PR TITLE
Full rewards script

### PR DESCRIPTION
**Pull request by cameron1024**
First attempt at the full rewards script, which accepts a config file given via `catalyst-toolbox rewards full path/to/config`.

I'm not sure all the inputs/outputs are right, in particular, a few of the inputs look duplicated, and there's no obvious voter output (it seems to just write to the block0 file :/ )

### CODE REVIEW

File: catalyst-toolbox/src/bin/cli/rewards/full/config.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    pub(super) active_voteplans: PathBuf,
+    pub(super) challenges: PathBuf,
+    pub(super) proposals_path: PathBuf,
+}
+
+#[derive(Debug, Deserialize)]
+pub(super) struct Outputs {
+    pub(super) veterans_rewards_output: PathBuf,
+    pub(super) ca_rewards_output: PathBuf,
+    pub(super) proposer_rewards_output: PathBuf,
```
Comment (Zeegomo) [inline catalyst-toolbox/src/bin/cli/rewards/full/config.rs:34]: voter rewards are missing

File: catalyst-toolbox/src/bin/cli/rewards/full/config.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+#[derive(Debug, Deserialize)]
+pub(super) struct Params {
+    pub(super) registration_threshold: u64,
+    pub(super) vote_threshold: u64,
+    pub(super) total_rewards: u64,
```
Comment (Zeegomo) [inline catalyst-toolbox/src/bin/cli/rewards/full/config.rs:41]: I think this should be separated for each rewards type

File: catalyst-toolbox/src/bin/cli/rewards/full/proposers.rs [hunk 3]
```diff
@@ -0,0 +1,37 @@
+use std::{ffi::OsStr, path::Path};
+
+use color_eyre::Result;
+
+use super::python::exec_python_script;
+
+#[allow(clippy::too_many_arguments)]
+pub(super) fn proposers_rewards(
```
Comment (Zeegomo) [inline catalyst-toolbox/src/bin/cli/rewards/full/proposers.rs:8]: IIRC, we also need to massage data before this step with `csv_merger.py`

File: catalyst-toolbox/src/bin/cli/rewards/voters.rs [hunk 4]
```diff
@@ -65,30 +68,51 @@ impl VotersRewards {
...
+        vote_threshold,
+        &block0,
+        snapshot,
+        Rewards::from(total_rewards),
+    )?;
+
+    let actual_rewards = results.values().sum::<Rewards>();
+    assert_are_close(actual_rewards, Rewards::from(total_rewards));
+
+    write_rewards_results(&Some(block_file.to_path_buf()), &results)?;
```
Comment (Zeegomo) [inline catalyst-toolbox/src/bin/cli/rewards/voters.rs:116]: why writing to block0? this should have its own output file

**Review by Zeegomo — CHANGES_REQUESTED** [review]

File: catalyst-toolbox/src/bin/cli/rewards/voters.rs (hunk 4 above)
Comment (cameron1024) [reply catalyst-toolbox/src/bin/cli/rewards/voters.rs:116]: I got mixed up when splitting this code out, I didn't see that it was using `common.output`

File: catalyst-toolbox/src/bin/cli/rewards/full/config.rs (hunk 2 above)
Comment (cameron1024) [reply catalyst-toolbox/src/bin/cli/rewards/full/config.rs:41]: Is this referring to the conversation in the ticket? I read through it but wasn't sure what the implications were for this PR. Is the "reward type" refrerring to ada/lovelace/dollars or something else?

Comment (Zeegomo) [reply catalyst-toolbox/src/bin/cli/rewards/full/config.rs:41]: Nono, I mean that each rewards (i.e. proposer, voters, vca, ca) will likely have a different value for `total_rewards`, e.g. 100 ADA for voter rewards, 50 for vCA

**Comment by cameron1024** [comment]
This is currently failing CI due to a patch for jormungandr crates, once https://github.com/input-output-hk/jormungandr/pull/4046 is merged, we can remove these patches and get this PR green

**Comment by minikin** [comment]
@cameron1024 Can we make this PR green?

File: catalyst-toolbox/src/rewards/proposers/util.rs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    path.push(&*challenge_name);
+    let path = PathBuf::from(path);
+
+    match ext {
+        Some(ext) => path.with_extension(ext),
+        None => path,
+    }
+}
+
+fn sanitize_name(name: &str) -> Cow<'_, str> {
```
Comment (cong-or) [inline catalyst-toolbox/src/rewards/proposers/util.rs:24]: https://github.com/input-output-hk/catalyst-toolbox/pull/146/files#diff-0b91cc9037e52e3f47667ccf7a7f8f61a6f27a42f33a854c6adb3de87568706aR24 seems like replication? is the other PR still relevant?

Comment (cameron1024) [reply catalyst-toolbox/src/rewards/proposers/util.rs:24]: Huh, not quite sure what's happened with these PRs. This one must have got mixed up with the rust rewrite at some point :shrug:

File: catalyst-toolbox/src/bin/cli/rewards/veterans.rs [hunk 6]
```diff
@@ -78,47 +78,75 @@ impl VeteransRewards {
             reputation_agreement_rate_cutoffs,
             reputation_agreement_rate_modifiers,
         } = self;
-        let reviews: Vec<VeteranRankingRow> = csv::load_data_from_csv::<_, b','>(&from)?;
 
-        if rewards_agreement_rate_cutoffs.len() != rewards_agreement_rate_modifiers.len() {
-            bail!(
+        vca_rewards(
```
Comment (cong-or) [inline catalyst-toolbox/src/bin/cli/rewards/veterans.rs:82]: any tests for the proposer logic?

File: catalyst-toolbox/src/rewards/proposers/types.rs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+pub enum NotFundedReason {
+    #[serde(rename = "Not Funded - Over Budget")]
+    OverBudget,
+    #[serde(rename = "Not Funded - Approval Threshold")]
+    ApprovalThreshold,
+}
+
+#[derive(StructOpt)]
+#[structopt(rename_all = "kebab-case")]
```
Comment (cong-or) [inline catalyst-toolbox/src/rewards/proposers/types.rs:40]: clap would be nice in the future.

File: catalyst-toolbox/src/bin/cli/rewards/full/config.rs [hunk 8]
```diff
@@ -0,0 +1,76 @@
+use std::path::PathBuf;
```
Comment (cong-or) [inline catalyst-toolbox/src/bin/cli/rewards/full/config.rs:1]: are votes rewards in here?

File: catalyst-toolbox/src/bin/cli/rewards/full/config.rs (hunk 2 above)
Comment (Mr-Leshiy) [reply catalyst-toolbox/src/bin/cli/rewards/full/config.rs:68]: @cameron1024  Are we going to split this rewards in this PR ? Or it should be a part of another ticket ?

File: catalyst-toolbox/src/rewards/proposers/types.rs (hunk 7 above)
Comment (cameron1024) [reply catalyst-toolbox/src/rewards/proposers/types.rs:40]: Yep, we're stuck on a version before the derive API was added IIRC. But clap 4 just got released :/

File: catalyst-toolbox/src/bin/cli/rewards/veterans.rs (hunk 6 above)
Comment (cameron1024) [reply catalyst-toolbox/src/bin/cli/rewards/veterans.rs:82]: This stuff is covered by the vit-testing PR

File: catalyst-toolbox/src/bin/cli/rewards/full/config.rs (hunk 8 above)
Comment (cameron1024) [reply catalyst-toolbox/src/bin/cli/rewards/full/config.rs:1]: Yep, there's a few places (e.g. `vote_count_path` on line 19)

**Review by cong-or — APPROVED** [review]

**Pull request merged**